tests/unit_tests/test_matchmaker_queue_algorithm.py

```diff
@@ -439,3 +445,129 @@ def test_make_matches_communicates_failed_attempts(p):
+    teams to be formed by putting players with the same rating on the same team.
+    """
+
+    # Large enough so the test is unlikely to pass by chance
+    num = 40
+
+    searches = [Search([p(1000, 10, name=f"p{i}")]) for i in range(num)]
+    searches += [Search([p(500, 10, name=f"p{i}")]) for i in range(num)]
+    matched, non_matched = algorithm.make_teams_from_single(searches, size=2)
+
+    assert matched != []
+    assert non_matched == []
+
+    for search in matched:
+        p1, p2 = search.players
+        assert p1.ratings[RatingType.LADDER_1V1] == p2.ratings[RatingType.LADDER_1V1]
+
+
+def test_make_teams_single_2v2_small_pool(p):
+    """
+    When we have a small number of players, we want teams to be formed by
+    distributing players of equal skill to different teams so that we can
+    maximize the chances of getting a match.
+    """
+
+    # Try a bunch of times so it is unlikely to pass by chance
+    for _ in range(20):
+        searches = [Search([p(1000, 10, name=f"p{i}")]) for i in range(2)]
+        searches += [Search([p(500, 10, name=f"r{i}")]) for i in range(2)]
+        matched, non_matched = algorithm.make_teams_from_single(searches, size=2)
+
+        assert matched != []
+        assert non_matched == []
+
+        for search in matched:
+            p1, p2 = search.players
+            # Order doesn't matter
+            if p1.ratings[RatingType.LADDER_1V1] == (1000, 10):
+                assert p2.ratings[RatingType.LADDER_1V1] == (500, 10)
+            else:
+                assert p1.ratings[RatingType.LADDER_1V1] == (500, 10)
+                assert p2.ratings[RatingType.LADDER_1V1] == (1000, 10)
+
+
+def test_make_buckets_performance(bench, p):
+    NUM_SEARCHES = 1000
+    searches = [Search([p(1500, 500, ladder_games=1)]) for _ in range(NUM_SEARCHES)]
+
+    with bench:
+        algorithm._make_buckets(searches)
+
+    assert bench.elapsed() < 0.1
+
+
+def test_make_teams_1(p):
+    teams = [
+        [p(name="p1"), p(name="p2"), p(name="p3")],
+        [p(name="p4"), p(name="p5")],
+        [p(name="p6"), p(name="p7")],
+        [p(name="p8")],
+        [p(name="p9")],
+        [p(name="p10")],
+        [p(name="p11")]
+    ]
+    do_test_make_teams(teams, 3, 2, {1})
+
+
+def test_make_teams_2(p):
+    teams = [
+        [p(name="p1"), p(name="p2"), p(name="p3")],
+        [p(name="p4"), p(name="p5")],
+        [p(name="p6"), p(name="p7")],
+        [p(name="p8")],
+        [p(name="p9")],
+        [p(name="p10")],
+        [p(name="p11")]
+    ]
+    do_test_make_teams(teams, 2, 1, {3})
+
+
+def test_make_teams_3(p):
+    teams = [
+        [p(name=f"p{i+1}")] for i in range(9)
+    ]
+    do_test_make_teams(teams, 4, 1, {1})
+
+
+def test_make_teams_4(p):
+    teams = [[p()] for i in range(9)]
+    teams += [[p(), p()] for i in range(5)]
+    teams += [[p(), p(), p()] for i in range(15)]
+    teams += [[p(), p(), p(), p()] for i in range(4)]
+    do_test_make_teams(teams, 4, 7, {3, 2})
```

Comment:
Maybe this could also for example assert that there is at most one unmatched team of size 2,
otherwise two unmatched teams of size 2 could have been combined.